examples/jenkins-image-sample.groovy

```diff
@@ -0,0 +1,119 @@
+    def saSelector4 = openshift.selector( "dc", [ "template" : "mytemplate" ] )
+    
+    def templateSelector = openshift.selector( "template", "mongodb-ephemeral")
+        
+    templateSelector.delete('--ignore-not-found')
+    openshift.delete('all,secret', '-l', 'template=mongodb-ephemeral-template', '--ignore-not-found')
+    template = openshift.create('https://raw.githubusercontent.com/openshift/origin/master/examples/db-templates/mongodb-ephemeral-template.json').object()
+    
+    // Explore the Groovy object which models the OpenShift template as a Map
+    echo "Template contains ${template.parameters.size()} parameters"
+
+    // For fun, modify the template easily while modeled in Groovy
+    template.labels["mylabel"] = "myvalue"
+
+    // Process the modeled template. We could also pass JSON/YAML or a template name instead.
+    // note: -p option for oc process not in the oc version that we currently ship with openshift jenkins images
+    def objectModels = openshift.process( template )//, "-p", "MEMORY_LIMIT=600Mi")
+
+    // objectModels is a list of objects the template defined, modeled as Groovy objects
+    echo "The template will create ${objectModels.size()} objects"
+
+    // For fun, modify the objects that have been defined by processing the template
+    for ( o in objectModels ) {
+        o.metadata.labels[ "anotherlabel" ] = "anothervalue"
+    }
+
+    // Serialize the objects and pass them to the create API.
+    // We could also pass JSON/YAML directly; openshift.create(readFile('some.json'))
+    def created = openshift.create(objectModels)
+
+    // Create returns a selector which will always select the objects created
+    created.withEach {
+        // Each loop binds the variable 'it' to a selector which selects a single object
+        echo "Created ${it.name()} from template with labels ${it.object().metadata.labels}"
+    }
+
+    // Filter created objects and create a selector which selects only the new DeploymentConfigs
+    def dcs = created.narrow("dc")
+    echo "Database will run in deployment config: ${dcs.name()}"
+    timeout(5) {
+        // Find a least one pod related to the DeploymentConfig and wait it satisfies a condition
+        dcs.related('pods').untilEach(1) {
+            // untilEach only terminates when each selected item causes the body to return true
+            return it.object().status.phase != 'Pending'
+        }
+    }
+
+    // Print out all pods created by the DC
+    echo "Template created pods: ${dcs.related('pods').names()}"
+
+    // Show how we can use labels to select as well
+    echo "Finding dc using labels instead: ${openshift.selector('dc',[mylabel:'myvalue']).names()}"
+
+    echo "DeploymentConfig description"
+    dcs.describe()
+    echo "DeploymentConfig history"
+    dcs.rollout().history()
+
+    def rubySelector = openshift.selector("bc", "ruby")
+    def builds
+    try {
+        rubySelector.object()
```

Comment:
Is it possible to handle this without catching exceptions? Since this is educational, I'd like it to propose a best practice.
```
def builds = rubySelector.related('builds') // This can be created even if bc/ruby doesn't exist yet
if ( rubySelector.count() == 0 ) { 
...newBuild...
}
```

You might even consider eliminating reuse altogether in this sample by deleting the bc/builds. The comment at the top says the test cleans up after itself.